File: CLI-gpt/cli_gpt/selectors.py
"""ChatGPT-specific DOM selectors, kept out of browser orchestration code."""

from __future__ import annotations

import re
from collections.abc import Iterable
from typing import Any

from .errors import PromptBoxNotFound


PROMPT_NAME = re.compile(r"message|prompt|chatgpt|메시지|프롬프트", re.IGNORECASE)
SEND_NAME = re.compile(r"send(?: prompt| message)?|보내기|전송", re.IGNORECASE)
STOP_NAME = re.compile(r"stop (?:generating|streaming)|생성 중지|응답 중지", re.IGNORECASE)
NEW_CHAT_NAME = re.compile(r"new chat(?: in .+)?|새 채팅", re.IGNORECASE)
LOGIN_NAME = re.compile(r"log ?in|sign ?in|로그인", re.IGNORECASE)


# Project-reading selectors stay in this single module so UI repairs do not leak
# into discovery, storage, or controller code. Links are additionally validated
# against the requested project id before they are accepted.
PROJECT_SPECIFIC_CONVERSATION_REGIONS = (
    '[data-testid="project-conversations"]',
    '[data-testid*="project" i][data-testid*="conversation" i]',
)
PROJECT_CONVERSATION_REGIONS = (
    *PROJECT_SPECIFIC_CONVERSATION_REGIONS,
    'main [role="list"]',
    "main",
)
PROJECT_CHAT_LINKS = ('a[href*="/c/"]',)
PROJECT_SPECIFIC_NAMES = (
    '[data-testid="project-name"]',
    '[aria-label*="Project" i] h1',
)
PROJECT_NAMES = (
    *PROJECT_SPECIFIC_NAMES,
    "main h1",
)
PROJECT_EMPTY_STATES = (
    '[data-testid*="empty" i]',
    '[data-testid*="no-conversation" i]',
)
PROJECT_EMPTY_NAME = re.compile(
    r"no (?:chats|conversations)|start (?:a |your )?(?:chat|conversation)"
    r"|대화가 없습니다|채팅이 없습니다|새 채팅",
    re.IGNORECASE,
)
CONVERSATION_TITLES = (
    '[data-testid="conversation-title"]',
    'nav [aria-current="page"]',
    "main h1",
)
CONVERSATION_EMPTY_STATES = (
    '[data-testid="conversation-empty-state"]',
    '[data-testid*="empty-conversation" i]',
)
MESSAGE_ROOTS = (
    '[data-message-author-role="user"]',
    '[data-message-author-role="assistant"]',
    'article[data-testid^="conversation-turn"] [data-message-author-role]',
)
MESSAGE_UI_EXCLUSIONS = (
    "button",
    "svg",
    '[data-testid*="copy" i]',
    '[data-testid*="feedback" i]',
    '[data-testid*="reaction" i]',
    '[aria-hidden="true"]',
)
PROJECT_ACCESS_ERROR_NAME = re.compile(
    r"not found|no access|do not have access|permission"
    r"|찾을 수 없|접근.*없|권한",
    re.IGNORECASE,
)


def _each_match(locator: Any) -> Iterable[Any]:
    try:
        count = locator.count()
    except Exception:
        return
    for index in range(count):
        try:
            yield locator.nth(index)
        except Exception:
            continue


def _usable(locator: Any, *, editable: bool = False) -> bool:
    try:
        if not locator.is_visible() or not locator.is_enabled():
            return False
        return not editable or locator.is_editable()
    except Exception:
        return False


def _first_usable(candidates: Iterable[Any], *, editable: bool = False):
    for candidate in candidates:
        for locator in _each_match(candidate):
            if _usable(locator, editable=editable):
                return locator
    return None


def prompt_box_candidates(page: Any) -> list[Any]:
    return [
        page.get_by_role("textbox", name=PROMPT_NAME),
        page.locator('[aria-label="Prompt"], [aria-label="Message ChatGPT"], [aria-label="메시지"]'),
        page.locator("#prompt-textarea"),
        page.locator('[data-testid="prompt-textarea"]'),
        page.locator('div[contenteditable="true"][role="textbox"]'),
        page.locator('div[contenteditable="true"]'),
        page.locator("textarea"),
    ]


def find_prompt_box(page: Any):
    locator = _first_usable(prompt_box_candidates(page), editable=True)
    if locator is None:
        raise PromptBoxNotFound(
            "Could not find an enabled ChatGPT prompt box. The page may require login or its UI may have changed."
        )
    return locator


def find_send_button(page: Any):
    return _first_usable(
        [
            page.get_by_role("button", name=SEND_NAME),
            page.locator('[data-testid="send-button"]'),
            page.locator('button[aria-label*="Send" i], button[aria-label*="보내기"]'),
        ]
    )


def find_stop_button(page: Any):
    return _first_usable(
        [
            page.get_by_role("button", name=STOP_NAME),
            page.locator('[data-testid="stop-button"]'),
            page.locator('button[aria-label*="Stop" i], button[aria-label*="중지"]'),
        ]
    )


def find_new_chat_control(page: Any):
    return _first_usable(
        [
            page.get_by_role("button", name=NEW_CHAT_NAME),
            page.get_by_role("link", name=NEW_CHAT_NAME),
            page.locator('[data-testid="create-new-chat-button"]'),
        ]
    )


def login_or_challenge_visible(page: Any) -> bool:
    candidates = [
        page.get_by_role("button", name=LOGIN_NAME),
        page.get_by_role("link", name=LOGIN_NAME),
        page.locator('iframe[src*="captcha" i], iframe[src*="challenge" i]'),
        page.locator('[id*="captcha" i], [class*="captcha" i]'),
    ]
    return _first_usable(candidates) is not None


def project_access_error_visible(page: Any) -> bool:
    return (
        _first_usable(
            [
                page.get_by_role("heading", name=PROJECT_ACCESS_ERROR_NAME),
                page.get_by_role("alert", name=PROJECT_ACCESS_ERROR_NAME),
                page.get_by_text(PROJECT_ACCESS_ERROR_NAME),
            ]
        )
        is not None
    )


def assistant_response_count(page: Any) -> int:
    counts: list[int] = []
    for selector in (
        '[data-message-author-role="assistant"]',
        'article[data-testid^="conversation-turn"] [data-message-author-role="assistant"]',
    ):
        try:
            counts.append(page.locator(selector).count())
        except Exception:
            continue
    return max(counts, default=0)


def latest_assistant_fingerprint(page: Any) -> str:
    locator = page.locator('[data-message-author-role="assistant"]')
    try:
        count = locator.count()
        if not count:
            return ""
        text = locator.nth(count - 1).inner_text(timeout=500)
        return f"{count}:{len(text)}:{text[-80:]}"
    except Exception:
        return ""

File: extension/content/selectors.js
export const SELECTORS = Object.freeze({
  stopControls: [
    'button[data-testid="stop-button"]',
    'button[aria-label="Stop generating"]',
    'button[aria-label="Stop streaming"]',
    'button[aria-label*="생성 중지"]',
    'button[aria-label*="응답 중지"]'
  ],
  messageRoots: [
    '[data-message-author-role="user"]',
    '[data-message-author-role="assistant"]'
  ],
  conversationTitles: [
    '[data-testid="conversation-title"]',
    'nav [aria-current="page"]',
    'main h1'
  ],
  projectNames: [
    '[data-testid="project-name"]',
    '[aria-label*="Project"] h1',
    'main h1'
  ]
});

function isVisible(element) {
  if (!(element instanceof Element)) return false;
  const style = getComputedStyle(element);
  const rectangle = element.getBoundingClientRect();
  return style.visibility !== "hidden" && style.display !== "none" && rectangle.width > 0 && rectangle.height > 0;
}

export function firstVisible(selectors, root = document) {
  for (const selector of selectors) {
    let elements;
    try {
      elements = root.querySelectorAll(selector);
    } catch (error) {
      console.warn(`[outoGPT] Invalid selector: ${selector}`, error);
      continue;
    }
    for (const element of elements) {
      if (isVisible(element)) return { element, selector };
    }
  }
  return null;
}

export function findStopControl(root = document) {
  const semantic = firstVisible(SELECTORS.stopControls, root);
  if (semantic) return semantic;

  const accessiblePattern = /^(stop (generating|streaming)|생성 중지|응답 중지)$/i;
  for (const button of root.querySelectorAll('button, [role="button"]')) {
    const name = (button.getAttribute("aria-label") || button.textContent || "").trim();
    if (accessiblePattern.test(name) && isVisible(button)) {
      return { element: button, selector: "accessible-name-fallback" };
    }
  }
  return null;
}

export function hasStopControl(root = document) {
  return findStopControl(root) !== null;
}

export function findMessageElements(root = document) {
  const primary = [...root.querySelectorAll('[data-message-author-role="user"], [data-message-author-role="assistant"]')];
  if (primary.length) return primary;
  return [...root.querySelectorAll('article[data-testid^="conversation-turn"]')]
    .filter((element) => element.querySelector('[data-message-author-role="user"], [data-message-author-role="assistant"]'));
}

export function getRoleForMessage(element) {
  const roleNode = element.matches?.("[data-message-author-role]")
    ? element
    : element.querySelector?.("[data-message-author-role]");
  const role = roleNode?.getAttribute("data-message-author-role");
  return role === "user" || role === "assistant" ? role : null;
}

export function findConversationTitle(root = document) {
  const match = firstVisible(SELECTORS.conversationTitles, root);
  const text = match?.element?.textContent?.trim();
  if (text && text.length <= 300) return text;
  const documentTitle = root.title?.replace(/\s*[|–-]\s*ChatGPT\s*$/i, "").trim();
  return documentTitle && documentTitle.toLowerCase() !== "chatgpt" ? documentTitle : null;
}

export function findProjectNameCandidate(root = document) {
  const match = firstVisible(SELECTORS.projectNames, root);
  const text = match?.element?.textContent?.trim();
  if (text && text.length <= 200 && !/^chatgpt$/i.test(text)) {
    return { name: text, selector: match.selector };
  }
  const documentTitle = root.title?.replace(/\s*[|–-]\s*ChatGPT\s*$/i, "").trim();
  if (documentTitle && !/^chatgpt$/i.test(documentTitle)) {
    return { name: documentTitle, selector: "document.title" };
  }
  return null;
}

